xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java

```diff
@@ -0,0 +1,169 @@
+      @SuppressWarnings("unchecked")
+      @Override
+      public void run() {
+        HashMap<String, Object> copyResources = (HashMap<String, Object>) new HashMap<>(resources);
+        xdsResources.put(type, copyResources);
+        for (Map.Entry<StreamObserver<DiscoveryResponse>, List<String>> entry :
+            subscribers.get(type).entrySet()) {
+          entry.getKey().onNext(generateResponse(type, entry.getValue()));
+        }
+      }
+    });
+  }
+
+  @Override
+  public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+      final StreamObserver<DiscoveryResponse> responseObserver) {
+    final StreamObserver<DiscoveryRequest> requestObserver =
+        new StreamObserver<DiscoveryRequest>() {
+      @Override
+      public void onNext(final DiscoveryRequest value) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            logger.log(Level.FINEST, "control plane received request {0}", value);
+            if (value.hasErrorDetail()) {
+              logger.log(Level.FINE, "control plane received nack resource {0}, error {1}",
+                  new Object[]{value.getResourceNamesList(), value.getErrorDetail()});
+              return;
+            }
+            if (value.getResourceNamesCount() <= 0) {
+              return;
+            }
+            String resourceType = value.getTypeUrl();
+            if (!value.getResponseNonce().isEmpty()
+                && !String.valueOf(xdsNonces.get(resourceType)).equals(value.getResponseNonce())) {
+              logger.log(Level.FINE, "Resource nonce does not match, ignore.");
+              return;
+            }
+            if (String.valueOf(xdsVersions.get(resourceType)).equals(value.getVersionInfo())) {
+              logger.log(Level.FINEST, "control plane received ack for resource: {0}",
+                  value.getResourceNamesList());
+              return;
+            }
+            responseObserver.onNext(
+                generateResponse(resourceType, value.getResourceNamesList()));
+            subscribers.get(resourceType)
+                .put(responseObserver, new ArrayList<>(value.getResourceNamesList()));
+          }
+        });
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        logger.log(Level.FINE, "Control plane error: {0} ", t);
+      }
+
+      @Override
+      public void onCompleted() {
+        responseObserver.onCompleted();
+        for (String type : subscribers.keySet()) {
+          subscribers.get(type).remove(responseObserver);
+        }
+      }
+    };
+    return requestObserver;
+  }
+
+  //must run in syncContext
+  private DiscoveryResponse generateResponse(String resourceType, List<String> resourceNames) {
+    DiscoveryResponse.Builder responseBuilder = DiscoveryResponse.newBuilder()
+        .setTypeUrl(resourceType)
+        .setVersionInfo(String.valueOf(xdsVersions.get(resourceType).getAndDecrement()))
```

Comment:
The version should change each time `setXdsConfig()` is called, not each time a response is generated. Right? Seems the `getAndDecrement()` should just move to `setXdsConfig()`.